gpkit/constraints/signomial_program.py

```diff
@@ -132,48 +109,18 @@ def localsolve(self, solver=None, verbosity=1, x0=None, rel_tol=1e-4,
         # solved successfully!
         if verbosity > 0:
             print("Solving took %i GP solves" % len(self.gps)
-                  + " and %.3g seconds." % (time() - self.starttime))
-
-        # parse the result and return nu's of original monomials from
-        #  variable sensitivities
-        nu = result["sensitivities"]["monomials"]
-        sens_vars = {var: sum([gp.exps[i][var]*nu[i] for i in locs])
-                     for (var, locs) in gp.varlocs.items()}
-        nu_ = []
-        for signomial in self.signomials:
-            for c, exp in zip(signomial.cs, signomial.exps):
-                var_ss = [sens_vars[var]*val for var, val in exp.items()]
-                nu_.append(functools_reduce(mul, var_ss, np.sign(c)))
-        result["sensitivities"]["monomials"] = np.array(nu_)
-        # TODO: SP sensitivities are weird, and potentially incorrect
+                  + " and %.3g seconds." % (time() - starttime))
 
+        result["signomialstart"] = startpoint
+        self.sens_from_gpconstr(gp.constraints,
+                                result["sensitivities"]["constraints"],
+                                result["sensitivities"]["constants"])
         self.result = result  # NOTE: SIDE EFFECTS
         return result
 
-    def step(self, x0=None, verbosity=1):
-        if x0 is None:
-            # dummy nomial data to turn x0's keys into VarKeys
-            self.negydata = lambda: None
-            self.negydata.varlocs = self.negvarkeys
-            self.negydata.varstrs = {str(vk): vk for vk in self.negvarkeys}
-            x0 = get_constants(self.negydata, {})
-            sp_inits = {vk: vk.descr["sp_init"] for vk in self.negvarkeys
-                        if "sp_init" in vk.descr}
-            x0.update(sp_inits)
-        # HACK: initial guess for negative variables
-        x0.update({var: 1 for var in self.negvarkeys if var not in x0})
-
-        posy_approxs = []
-        for p, n in zip(self.posynomials, self.negynomials):
-            if n is 0:
-                posy_approx = p
-            else:
-                posy_approx = p/n.mono_lower_bound(x0)
-            posy_approxs.append(posy_approx)
-
-        gp = GeometricProgram(posy_approxs[0], posy_approxs[1:],
-                              verbosity=verbosity)
-        return gp
+    def gp(self, x0=None, verbosity=1):
```

Comment:
Note that this is partly repeated in Model, perhaps calling for ConstrainBase or CostedConstraintSet to be the one that implements `gp`